### PR TITLE
mimxrt: Add initial PSRAM implementation.

### LINKED ISSUE #18281: mimxrt: PSRAM support for Teensy 4.1
### Description

PSRAM support for Teensy 4.1! And other applicable boards 😉

Teensy 4.1 has pads for PSRAM chips to be added, but I was unable to find any mention of PSRAM in the entire `mimxrt` port. Nor could I find any discussions/issues/PRs, so figured I'd put in the request!

### Code Size

_No response_

### Implementation

I hope the MicroPython maintainers or community will implement this feature

### Code of Conduct

Yes, I agree

### PR DESCRIPTION
**Pull request by sfe-SparkFro**
### Summary

PSRAM support for Teensy 4.1 and other `mimxrt` boards! Resolves #18281

This implementation is a copy of the [Arduino core PSRAM implementation here](https://github.com/PaulStoffregen/cores/blob/a23cade55c7530cce43c641ec3ddb8b7f83148a8/teensy4/startup.c#L379-L587). I modified the implementation to use the `#define`s from [`lib/nxp_driver/sdk/devices/MIMXRT1062/MIMXRT1062.h`](https://github.com/micropython/nxp_driver/blob/master/sdk/devices/MIMXRT1062/MIMXRT1062.h) where possible instead of the huge [`imxrt.h`](https://github.com/PaulStoffregen/cores/blob/master/teensy4/imxrt.h). However there's still about a dozen that I had to copy over from `imxrt.h` that currently live in `psram.c`; happy to change/move these if there's a better solution!

The MicroPython side of things is a copy of the `rp2` PSRAM PR (#15620). I have not looked into whether `MICROPY_ALLOC_GC_STACK_SIZE` needs to be increased nor if any clock timings need to change, like was done in the `rp2` PR. There may be other edge cases that need to be addressed, but I'm new to Teensy/mimxrt, so don't know what to look for :slightly_smiling_face:

### Testing

The following test code prints the total amount of memory seen by the garbage collector, then performs a memory test with a small and large `bytearray` to demonstrate the split heap working.

```
import gc
import uctypes

gc.collect()
print("Initial GC free (MiB):", gc.mem_free() / 1024 / 1024)

def test(size):
    gc.collect()
    b = bytearray(size)
    print("Test size:", size, "location: 0x{:X}".format(uctypes.addressof(b)), end=' ')
    for i in range(size):
        b[i] = i & 0xFF
    for i in range(size):
        if b[i] != (i & 0xFF):
            print("Fail at", i)
            return
    print("Pass!")

test(10)
test(10_000_000)
```

Testing with a single 16MiB chip, the GC shows over 16MiB available. The small array is allocated in the `0x20000000` range (DTCM), and the large array is allocated in the `0x70000000` range (PSRAM). All behaves as I'd expect!

```
Initial GC free (MiB): 16.358551025390626
Test size: 10 location: 0x20204720 Pass!
Test size: 10000000 location: 0x7005DD00 Pass!
```

Note that `test(10_000_000)` takes ~25 seconds on my end.

### Trade-offs and Alternatives

There could be an argument for PSRAM support being enabled by default, at least for Teensy 4.1 (set `#define MICROPY_HW_ENABLE_PSRAM (1)` in `boards/TEENSY41/mpconfigboard.h`). This implementation auto-detects the presence and size of 1 or 2 PSRAM chips if they're soldered, meaning a user wouldn't need to flash different firmware.

Building on my end, enabling PSRAM increases the firmware size from 626992 to 628064 bytes (+1072 / +0.17%), so I'll defer to others whether it's worth enabling it by default. If not, might be nice to create a variant with PSRAM enabled so it's easy to get from the [Teensy 4.1 download page](https://micropython.org/download/TEENSY41/).

### CODE REVIEW

**Comment by sfe-SparkFro** [comment]
Oh also, this would need permission from @PaulStoffregen, the code I copied does not have a license. Would be ideal if it could have the MIT license or similar added!

**Comment by robert-hh** [comment]
Works on my Teensy 4.1 with an 8 MB RAM chip. Total memory reported by mem_free() is 8941792 after having imported the usual boot-up stuff. So it's internal and external RAM, You test script reports:
```
Initial GC free (MiB): 8.546279907226562
Test size: 10 location: 0x20204750 Pass!
Test size: 5000000 location: 0x7002EEC0 Pass!
```

> the code I copied does not have a license.

I found once a note by Paul Stoffregen telling that the PJRC code is licensed under MIT or MIT-Like conditions: https://forum.pjrc.com/index.php?threads/open-source-license-issues-when-using-teensy-products.29382/. But that statement is pretty old.

**Comment by robert-hh** [comment]
You can enable PSRAM by default for Teensy 4.1. If the chip is not present (or blocked), just internal RAM is used. Tested on my board.
About the formatting errors. You can fix them with uncrustify using e.g.
`uncrustify --no-backup -c ../../tools/uncrustify.cfg psram.c`
You may have to install uncrustify.

File: ports/mimxrt/psram.c [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+ * THE SOFTWARE.
+ */
+
+#include "py/mphal.h"
+
+#if MICROPY_HW_ENABLE_PSRAM
+
+// Copied from here:
+// https://github.com/PaulStoffregen/cores/blob/a23cade55c7530cce43c641ec3ddb8b7f83148a8/teensy4/avr/pgmspace.h#L31
+#define FLASHMEM __attribute__((section(".flashmem")))
```
Comment (robert-hh) [inline ports/mimxrt/psram.c]: There is no reference to .flashmem in the loader file. So this definition seems obsolete. All code goes to flash anyhow, unless told otherwise.

File: ports/mimxrt/psram.c [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+// These #defines are from here:
+// https://github.com/PaulStoffregen/cores/blob/a23cade55c7530cce43c641ec3ddb8b7f83148a8/teensy4/imxrt.h
+#define CCM_CCGR_ON				3
+#define FLEXSPI_MCR2_CLRLEARNPHASE(x)		((uint32_t)(x<<14))
+#define FLEXSPI_LUTKEY_VALUE			((uint32_t)0x5AF05AF0)
+#define FLEXSPI_LUT_INSTRUCTION(opcode, pads, operand) ((uint32_t)(\
+	(((opcode) & 0x3F) << 10) | (((pads) & 0x03) << 8) | ((operand) & 0xFF)))
+#define FLEXSPI_LUT_OPCODE_CMD_SDR		0x01
```
Comment (robert-hh) [inline ports/mimxrt/psram.c]: There is already flexspi_flash_config.h which defines these constants and the LUT entry constructor, maybe worded slightly different.  So maybe it's better to switch to these definitions. It's in the context of flash, but the definitions are for flexspi. If the name of the file matters, it could be unified.

Comment (robert-hh) [reply ports/mimxrt/psram.c]: On a second glance, all of these constants are already defined in the lib header file fsl_flexspi.h. https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/drivers/flexspi/fsl_flexspi.h
So the could & should be used by both this file here and the flexspi_flash.h. I can do the latter.

**Comment by sfe-SparkFro** [comment]
> Works on my Teensy 4.1 with an 8 MB RAM chip.

Great to hear!

> I found once a note by Paul Stoffregen telling that the PJRC code is licensed under MIT or MIT-Like conditions: https://forum.pjrc.com/index.php?threads/open-source-license-issues-when-using-teensy-products.29382/. But that statement is pretty old.

Paul just pushed https://github.com/PaulStoffregen/cores/commit/10025393e83ca9f4dc5646643a41cb2f32022ae4 which adds license header comments.

> You can enable PSRAM by default for Teensy 4.1.

> About the formatting errors. You can fix them with uncrustify

Done!

File: ports/mimxrt/psram.c (hunk 1 above)
Comment (sfe-SparkFro) [reply ports/mimxrt/psram.c]: Done in 12091d3

File: ports/mimxrt/psram.c (hunk 2 above)
Comment (sfe-SparkFro) [reply ports/mimxrt/psram.c]: Done in 12091d3

**Comment by sfe-SparkFro** [comment]
Curse these commit and formatting checks... Should be fixed with force push of 12091d3

File: ports/mimxrt/psram.c [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+#include "py/mphal.h"
+#include "fsl_flexspi.h"
+
+#if MICROPY_HW_ENABLE_PSRAM
+
+// These #defines are from here:
+// https://github.com/PaulStoffregen/cores/blob/10025393e83ca9f4dc5646643a41cb2f32022ae4/teensy4/imxrt.h
+#define CCM_CCGR_ON                             3
+#define FLEXSPI_MCR2_CLRLEARNPHASE(x)           ((uint32_t)(x << 14))
+#define FLEXSPI_LUTKEY_VALUE                    ((uint32_t)0x5AF05AF0)
```
Comment (robert-hh) [inline ports/mimxrt/psram.c]: All of these are defined in fsl_flexspi.h as well as :

```
#define FREQ_1MHz             (1000000UL)
#define FLEXSPI_DLLCR_DEFAULT (0x100UL)
#define FLEXSPI_LUT_KEY_VAL   (0x5AF05AF0UL)
```

**Comment by sfe-SparkFro** [comment]
I got a note from Paul:

> Might be worthwhile to look at the configure_cache() function in startup.c.
> 
> [These 2 lines](https://github.com/PaulStoffregen/cores/blob/10025393e83ca9f4dc5646643a41cb2f32022ae4/teensy4/startup.c#L381-L382) enable the M7's cache for the PSRAM:
> 
>         SCB_MPU_RBAR = 0x70000000 | REGION(i++); // FlexSPI2
>         SCB_MPU_RASR = MEM_CACHE_WBWA | READWRITE | NOEXEC | SIZE_32M;

I poked around the MicroPython repo, and I don't see any evidence of the cache being configured. Or at least, the memory addresses for `SCB_MPU_RBAR` and `SCB_MPU_RASR` (`0xE000ED9C` and `0xE000EDA0` respectively) don't appear anywhere in the MicroPython repo (except for the `cc3200` port, but that's irrelevant to `mimxrt`).

IMO this is possibly a topic for a different thread, since I suspect it would impact more than just this PR. Looks like there's some other steps required to configure the cache, and my severe lack of experience with Teensy/`mimxrt` makes me feel quite unqualified to make any decisions or implement anything around it. But wanted to at least pass this on as a potential area for a performance boost.

File: ports/mimxrt/psram.c [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    IOMUXC->SW_MUX_CTL_PAD[29] = 8 | 0x10;     // ALT1 = FLEXSPI2_A_DATA3
+
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_DQS_FA_SELECT_INPUT] = 1;     // GPIO_EMC_23 for Mode: ALT8, pg 986
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_IO_FA_BIT0_SELECT_INPUT] = 1;     // GPIO_EMC_26 for Mode: ALT8
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_IO_FA_BIT1_SELECT_INPUT] = 1;     // GPIO_EMC_27 for Mode: ALT8
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_IO_FA_BIT2_SELECT_INPUT] = 1;     // GPIO_EMC_28 for Mode: ALT8
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_IO_FA_BIT3_SELECT_INPUT] = 1;     // GPIO_EMC_29 for Mode: ALT8
+    IOMUXC->SELECT_INPUT_1[kIOMUXC_FLEXSPI2_IPP_IND_SCK_FA_SELECT_INPUT] = 1;     // GPIO_EMC_25 for Mode: ALT8
+
+    // turn on clock  (QSPI flash & PSRAM chips usually spec max clock 100 to 133 MHz)
```
Comment (robert-hh) [inline ports/mimxrt/psram.c]: The clock setup in the following lines down to  line 137 inc. can be made a little bit more readable using the NXP lib macros:
```
    // Set FLEXSPI2_PODF divide by 5.
    CLOCK_SetDiv(kCLOCK_Flexspi2Div, 4);
    // Set Flexspi2 clock source to PLL2, 528 MHz.
    CLOCK_SetMux(kCLOCK_Flexspi2Mux, 3);
    CLOCK_EnableClock(kCLOCK_FlexSpi2);
```
The various other combinations may be added as a comment:

File: ports/mimxrt/psram.c [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+ */
+
+#include "py/mphal.h"
+#include "fsl_flexspi.h"
+
+#if MICROPY_HW_ENABLE_PSRAM
+
+// These #defines are from here:
+// https://github.com/PaulStoffregen/cores/blob/10025393e83ca9f4dc5646643a41cb2f32022ae4/teensy4/imxrt.h
+#define CCM_CCGR_ON                             3
```
Comment (robert-hh) [inline ports/mimxrt/psram.c]: This #define is not used.

**Comment by robert-hh** [comment]
There is DCACHE and ICACHE enabled in the port, which must be disabled when writing to flash. I do not know where the regions are defined.

**Comment by sfe-SparkFro** [comment]
FWIW I added a second 16MB PSRAM chip to my Teensy 4.1, and `gc.mem_free()` reports ~32MB as expected!

```
>>> gc.mem_free()/1024/1024
31.747390747070314
```

And `test(30_000_000)` also succeeds after ~70 seconds:

```
Test size: 30000000 location: 0x700BBA00 Pass!
```

So this implementation works as expected with 2 PSRAM chips!

**Comment by robert-hh** [comment]
Very good. I cannot do that with my board, since the second slot is populated with a flash chip.

I added a few comment sto the code lately, you may have seen them. One for replacing the pin numbers by the symbolic name from the lib headers, the other to use lib macros for setting the clock in stead of register names. Both just for better readability of the code.

**Comment by sfe-SparkFro** [comment]
> I added a few comment sto the code lately, you may have seen them.

I did not see them, where are they?

**Comment by robert-hh** [comment]
For pdsram.c, line 100ff, 
`IOMUXC->SW_PAD_CTL_PAD[22] = 0x1B0F9;     // 100K pullup, strong drive, max speed, hys`
line 109ff
`    IOMUXC->SW_MUX_CTL_PAD[22] = 8 | 0x10;     // ALT1 = FLEXSPI2_A_SS1_B (Flash)`
 and 125ff
`// turn on clock  (QSPI flash & PSRAM chips usually spec max clock 100 to 133 MHz)`
, if you look into the changed files here in the PR. And a comment for line 35
`#define CCM_CCGR_ON                             3`
and for the License note. Paul Stiffregen had in his License which you mentioned above two bullet points, but the license here juste mentions one of them.

**Comment by sfe-SparkFro** [comment]
Should be addressed with 8f50a00, hoping I used the correct macros and functions. Please let me know if there's anything else!

**Comment by robert-hh** [comment]
Looks better. If it works it's right. With the call to `CLOCK_SetDiv()` I had use symbolic values for the second argument, which reflect the effective divider, so instead of 
`CLOCK_SetDiv(kCLOCK_Flexspi2Div, 5)` it is `CLOCK_SetDiv(kCLOCK_Flexspi2Div, kCLOCK_Flexspi2DivBy6)`. kCLOCK_Flexspi2DivBy6 is defined as `(5)`, etc....
There might be a way to use the lib functions for configuring the flexspi unit instead of direct register access of lines 153ff. But even if possible that would require setting up a sparsely documented data structure with the intended configuration, and that may be not better readable than the existing code.

**Comment by sfe-SparkFro** [comment]
> If it works it's right.

Yes, still works fine on my end!

> instead of `CLOCK_SetDiv(kCLOCK_Flexspi2Div, 5)` it is `CLOCK_SetDiv(kCLOCK_Flexspi2Div, kCLOCK_Flexspi2DivBy6)`.

I see those are defined in [`_clock_div_value`](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L839-L846), happy to swap that out!

Do you know if there's an equivalent enum for `CLOCK_SetMux()` to replace the `3`? I don't see a `_clock_mux_value` enum or anything.


The original implementation uses `CCM_CBCMR_FLEXSPI2_CLK_SEL(3)`, which according to [`MIMXRT1062.h`](https://github.com/micropython/nxp_driver/blob/master/sdk/devices/MIMXRT1062/MIMXRT1062.h) should be `derive clock from PLL2 (pll2_main_clk)`:

```c
/*! FLEXSPI2_CLK_SEL
 *  0b00..derive clock from PLL2 PFD2
 *  0b01..derive clock from PLL3 PFD0
 *  0b10..derive clock from PLL3 PFD1
 *  0b11..derive clock from PLL2 (pll2_main_clk)
 */
#define CCM_CBCMR_FLEXSPI2_CLK_SEL(x)            (((uint32_t)(((uint32_t)(x)) << CCM_CBCMR_FLEXSPI2_CLK_SEL_SHIFT)) & CCM_CBCMR_FLEXSPI2_CLK_SEL_MASK)
```

So I thought maybe [`kCLOCK_SysPllClk`](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L491), but that has a value of `0xE0`.

Note that [`CLOCK_SetMux`](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L1408-L1426) does the exact same bit shifting (once you work through the nested macros), so a value of `3` is required as the second parameter:

```c
static inline void CLOCK_SetMux(clock_mux_t mux, uint32_t value)
{
    uint32_t busyShift;

    busyShift               = (uint32_t)CCM_TUPLE_BUSY_SHIFT(mux);
    CCM_TUPLE_REG(CCM, mux) = (CCM_TUPLE_REG(CCM, mux) & (~CCM_TUPLE_MASK(mux))) |
                              (((uint32_t)((value) << CCM_TUPLE_SHIFT(mux))) & CCM_TUPLE_MASK(mux));
...
}
```

So should I just leave it as `CLOCK_SetMux(kCLOCK_Flexspi2Mux, 3)`?

> There might be a way to use the lib functions for configuring the flexspi unit instead of direct register access

Looks like the correct functions are [here](https://github.com/micropython/nxp_driver/blob/master/sdk/drivers/flexspi/fsl_flexspi.h) (eg. `FLEXSPI_Init()`). Will see about changing that to see if it looks more readable!

**Comment by robert-hh** [comment]
> Note that [CLOCK_SetMux](https://github.com/micropython/nxp_driver/blob/91b04b34a59f6d81661cec6f84611afe6330ce92/sdk/devices/MIMXRT1062/drivers/fsl_clock.h#L1408-L1426) does the exact same bit shifting (once you work through the nested macros), so a value of 3 is required as the second parameter:

Yes. I know.

> Do you know if there's an equivalent enum for CLOCK_SetMux()

I looked for these and did not find any. But just naming the clock source may be less helpful than defining names with the actual frequency for the MUX constants (or use an enum). Like:
```
#define FlexspiClk_396MHz     (0)
#define FlexspiClk_720 Mhz    (1)
#define FlexspiClk_664_62 MHz (2)
#define FlexspiClk_528 MHz    (3)
```
The respective clock tree is in the MIMXRT1063 Reference manual, figure 14-2, part 1. It's quite a beast.

**Comment by sfe-SparkFro** [comment]
Right! That took a bit more effort than expected, but I've refactored it to use the library function calls in favor of direct register access (see df5b315). IMO it's a lot more readable now. Tested on my end and all is still working as expected!

**Comment by robert-hh** [comment]
Yes, that looks like you spent quite a bit of time. Luckily you knew what the result must be.

The code looks better now, especially since all the settings are named in the flexspi config data structures, making it easier to compare is against the manual. Only the size penalty of 896 bytes is much larger than expected. But there is still much space left in the text segment.

Tested with my board and a 8M RAM chip. Works fine.

**Comment by dpgeorge** [comment]
This is a nice addition!

But unfortunately the code added from PJRC is not MIT.  The extra clause seems quite problematic:

> 2. If the Software is incorporated into a build system that allows
selection among a list of target devices, then similar target
devices manufactured by PJRC.COM must be included in the list of
target devices and selectable in the same manner.

The "build system" here is our makefile-based build, and the "selection among a list of target devices" is our `ports/mimxrt/boards` directory.  To comply with this clause we "must include" "similar target devices manufactured by PJRC.COM".

There's a bit of ambiguity here, because AFAIK PJRC no longer manufactures devices (it's done by SparkFun, as you will be well aware!) so technically this clause is moot.  But in good faith of the clause it sounds like we should add support for all mimxrt-based Teensy boards.

I think we need clarification on this point from @PaulStoffregen.  Ideally any code added in this repo is proper MIT license.

**Comment by robert-hh** [comment]
Two notes:
- As far as I can tell, there is only one MIMXRT based teensy board which supports PSRAM, and that is the 4.1. The 4.0 board does not have the footprint for PSRAM, and the other Teensy boards are not MIMXRT, they use MCU of the Kinetis series..
- The code in this PR diverged a lot from the initial implementation by PJRC. While the first version used the PJRC code, it was replaced by new code. PJRC used direct register accesses to setup the hardware, while the final code makes use of the library data structures and methods, like the existing flash driver. And obviously the integration into MicroPyhon startup sequence is different.

**Comment by PaulStoffregen** [comment]
To the extent any of this code is derived from my work, I hereby give permission to publish it with MIT license.

**Comment by dpgeorge** [comment]
Thank you @PaulStoffregen for your prompt reply here, and agreeing to license the code as MIT.  That makes it a lot easier for us!

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Squashed and merged in 0e3cc2910d250b51f2c22ef32c2be73f049696da
